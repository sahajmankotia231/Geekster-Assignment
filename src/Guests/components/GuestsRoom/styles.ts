import styled from "styled-components";

import { ReactComponent as CloseSvg } from "assets/close.svg";
import { LIGHT_BORDER_GRAY, HEAVY_BLACK, RED } from "constants/colors";

export const Container = styled.div`
  width: 100%;
  border-bottom: 1px solid ${LIGHT_BORDER_GRAY};
  padding-bottom: 24px;
  margin-bottom: 24px;
`;

export const TitleContainer = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin-bottom: 18px;
`;

export const Title = styled.h4`
  color: ${HEAVY_BLACK};
  font-size: 18px;
  font-weight: 600;
  line-height: 24px;
`;

export const Guest = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin-bottom: 16px;
`;

export const GuestText = styled.span`
  color: ${HEAVY_BLACK};
  font-size: 16px;
  font-weight: 600;
  line-height: 20px;
`;

export const GuestButtonGroup = styled.div`
  width: 130px;
  align-items: center;
  display: flex;
  justify-content: space-between;

  ${GuestText} {
    margin: 0 20px;
  }
`;

export const KidsContainer = styled.div`
  border-left: 1px solid ${LIGHT_BORDER_GRAY};
  display: block;
  margin-left: 8px;
  padding-left: 10px;
`;

export const IconRemoveChild = styled(CloseSvg)`
  color: ${RED};
  fill: currentColor;
`;
