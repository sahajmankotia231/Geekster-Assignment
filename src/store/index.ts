import { createSlice, configureStore } from "@reduxjs/toolkit";

export type ChildType = {
  age: number | null;
};

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: ChildType[];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    initializeStore: (state, action): void => {
      if (Boolean(action.payload)) {
        const splitedRooms = action.payload?.split("|") || [];
        const rooms = splitedRooms?.map(
          (room): RoomType => {
            const guests = room.split(":");
            const adultsCounter = Number(guests[0]);
            const children =
              guests[1]?.split(",")?.reduce((arr, age): ChildType[] => {
                arr.push({ age: Number(age) });
                return arr;
              }, []) || [];

            return {
              ...initialRoomsState,
              adultsCounter,
              children,
            };
          }
        );

        state.rooms = rooms;
      }
    },
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  initializeStore,
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
