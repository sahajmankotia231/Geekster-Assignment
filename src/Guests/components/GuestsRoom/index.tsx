import React, { ReactElement } from "react";
import { useSelector, useDispatch } from "react-redux";

import Button from "components/Button";
import { StoreStateType, removeRoom } from "store";

import GuestsRoomAdult from "./components/GuestsRoomAdult";
import GuestsRoomChildren from "./components/GuestsRoomChildren";
import { Container, Title, TitleContainer } from "./styles";

type GuestsRoomProps = {
  position: number;
};

const GuestsRoom = ({ position }: GuestsRoomProps): ReactElement => {
  const dispatch = useDispatch();
  const rooms = useSelector((state: StoreStateType) => state.rooms);
  const isRoomRemovable = rooms.length > 1;

  return (
    <Container data-testid="guests-room">
      <TitleContainer>
        <Title>Room {position + 1}</Title>
        {isRoomRemovable && (
          <Button
            onClick={() => dispatch(removeRoom(position))}
            theme="danger"
            outline
            block={false}
            dataTestId={`remove-room-${position}-button`}
          >
            Remove room
          </Button>
        )}
      </TitleContainer>
      <GuestsRoomAdult position={position} />
      <GuestsRoomChildren position={position} />
    </Container>
  );
};

export default GuestsRoom;
