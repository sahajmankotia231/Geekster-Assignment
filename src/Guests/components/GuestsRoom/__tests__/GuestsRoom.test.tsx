import React from "react";
import { Provider } from "react-redux";
import { render, screen } from "@testing-library/react";

import store, { addRoom, resetStore } from "store";

import GuestsRoom from "..";

describe("<GuestsRoom />", () => {
  beforeEach(() => store.dispatch(resetStore()));

  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <GuestsRoom position={0} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should display the remove room button if at least 2 rooms exists", () => {
    const { rerender } = render(
      <Provider store={store}>
        <GuestsRoom position={0} />
      </Provider>
    );

    expect(screen.queryByText("Remove room")).toBeNull();

    store.dispatch(addRoom());
    rerender(
      <Provider store={store}>
        <GuestsRoom position={0} />
      </Provider>
    );

    expect(screen.queryByText("Remove room")).toBeTruthy();
  });
});
