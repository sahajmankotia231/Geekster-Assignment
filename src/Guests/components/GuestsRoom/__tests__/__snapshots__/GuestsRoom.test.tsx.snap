// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuestsRoom /> should render properly 1`] = `
<div
  class="sc-gsDJrp dwyZHA"
  data-testid="guests-room"
>
  <div
    class="sc-dkPtyc iQYZgR"
  >
    <h4
      class="sc-hKwCoD cLeDZh"
    >
      Room 
      1
    </h4>
  </div>
  <div
    class="sc-eCImvq guCNbL"
  >
    <span
      class="sc-jRQAMF bgpgfc"
    >
      Adults
    </span>
    <div
      class="sc-gKckTs jduWvm"
    >
      <button
        class="sc-bdvvaa fzcnyp"
        data-testid="remove-adult-0-button"
        disabled=""
      >
        <svg>
          minus.svg
        </svg>
      </button>
      <span
        class="sc-jRQAMF bgpgfc"
        data-testid="adults-counter"
      >
        1
      </span>
      <button
        class="sc-bdvvaa fzcnyp"
        data-testid="add-adult-0-button"
      >
        <svg>
          plus.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    class="sc-eCImvq guCNbL"
  >
    <span
      class="sc-jRQAMF bgpgfc"
    >
      Children
    </span>
    <div
      class="sc-gKckTs jduWvm"
    >
      <button
        class="sc-bdvvaa fzcnyp"
        data-testid="remove-child-0-button"
        disabled=""
      >
        <svg>
          minus.svg
        </svg>
      </button>
      <span
        class="sc-jRQAMF bgpgfc"
        data-testid="children-counter"
      >
        0
      </span>
      <button
        class="sc-bdvvaa fzcnyp"
        data-testid="add-child-0-button"
      >
        <svg>
          plus.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    class="sc-iCfLBT bZPXmC"
  />
</div>
`;
